File: cases/03-poor-observability-and-useless-logs/java/app/Main.java
import com.sun.net.httpserver.HttpExchange;
import com.sun.net.httpserver.HttpServer;

import java.io.IOException;
import java.io.OutputStream;
import java.net.InetSocketAddress;
import java.net.URI;
import java.net.URLDecoder;
import java.nio.charset.StandardCharsets;
import java.time.Instant;
import java.util.ArrayDeque;
import java.util.Collections;
import java.util.Deque;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.UUID;
import java.util.concurrent.Executors;
import java.util.concurrent.atomic.LongAdder;

/**
 * Caso 03 — Observabilidad deficiente y logs inutiles (stack Java).
 *
 * Legacy: System.out.println sin correlation, sin contexto. Errores opacos.
 * Observable: log estructurado JSON con correlationId + ScopedValue-like via
 * ThreadLocal, mas /logs endpoint que devuelve los ultimos N logs estructurados.
 *
 * Primitiva Java distintiva:
 *   - ThreadLocal<RequestContext> como contexto de correlation (espejo del
 *     ScopedValue de JDK 21 que requiere preview flags). Cada handler arranca
 *     un contexto y lo limpia al final.
 *   - LongAdder para contadores por tipo de evento.
 */
public class Main {

    private static final String CASE_NAME = "03 - Observabilidad deficiente y logs inutiles";
    private static final String STACK = "Java 21";
    private static final int PORT = Integer.parseInt(System.getenv().getOrDefault("PORT", "8080"));
    private static final int MAX_LOG_ENTRIES = 200;

    private static final Deque<String> recentLogs = new ArrayDeque<>();
    private static final ThreadLocal<RequestContext> CTX = new ThreadLocal<>();

    private static final LongAdder legacyErrors = new LongAdder();
    private static final LongAdder observableErrors = new LongAdder();
    private static final LongAdder legacyRequests = new LongAdder();
    private static final LongAdder observableRequests = new LongAdder();

    public static void main(String[] args) throws Exception {
        HttpServer server = HttpServer.create(new InetSocketAddress(PORT), 0);
        server.createContext("/", Main::route);
        server.setExecutor(Executors.newCachedThreadPool());
        server.start();
        System.out.println("[case03-java] listening on " + PORT);
        Runtime.getRuntime().addShutdownHook(new Thread(() -> server.stop(0)));
    }

    private static void route(HttpExchange ex) throws IOException {
        URI uri = ex.getRequestURI();
        String path = uri.getPath();
        Map<String, String> q = queryParams(uri.getRawQuery());
        int status = 200;
        String body;

        try {
            switch (path) {
                case "/":
                case "/index":
                    body = "{\"case\":\"" + CASE_NAME + "\",\"stack\":\"" + STACK +
                            "\",\"routes\":[\"/health\",\"/checkout-legacy?total=100\",\"/checkout-observable?total=100\",\"/logs\",\"/metrics\",\"/diagnostics/summary\",\"/reset-lab\"]}";
                    break;
                case "/health":
                    body = "{\"status\":\"ok\",\"stack\":\"" + STACK + "\",\"case\":\"" + CASE_NAME + "\"}";
                    break;
                case "/checkout-legacy":
                    body = checkoutLegacy(q.getOrDefault("total", "100"));
                    legacyRequests.increment();
                    break;
                case "/checkout-observable":
                    body = checkoutObservable(q.getOrDefault("total", "100"));
                    observableRequests.increment();
                    break;
                case "/logs":
                    body = logsJson();
                    break;
                case "/metrics":
                case "/diagnostics/summary":
                    body = diagnosticsJson();
                    break;
                case "/reset-lab":
                    legacyErrors.reset(); observableErrors.reset();
                    legacyRequests.reset(); observableRequests.reset();
                    synchronized (recentLogs) { recentLogs.clear(); }
                    body = "{\"status\":\"reset\"}";
                    break;
                default:
                    status = 404;
                    body = "{\"error\":\"not_found\",\"path\":\"" + escape(path) + "\"}";
            }
        } catch (Exception e) {
            status = 500;
            body = "{\"error\":\"internal\",\"detail\":\"" + escape(e.getMessage()) + "\"}";
        }

        byte[] out = body.getBytes(StandardCharsets.UTF_8);
        ex.getResponseHeaders().add("Content-Type", "application/json; charset=utf-8");
        ex.sendResponseHeaders(status, out.length);
        try (OutputStream os = ex.getResponseBody()) { os.write(out); }
    }

    /** Legacy: log opaco, sin correlation, sin ID, sin contexto. */
    private static String checkoutLegacy(String totalRaw) {
        double total = parseDoubleOr(totalRaw, 100.0);
        // log inutil
        System.out.println("[INFO] processing checkout");
        if (total > 500) {
            System.out.println("[ERROR] checkout failed");
            legacyErrors.increment();
            return "{\"variant\":\"legacy\",\"status\":\"error\",\"note\":\"log dice 'checkout failed' sin id, sin total, sin causa.\"}";
        }
        System.out.println("[INFO] checkout ok");
        return "{\"variant\":\"legacy\",\"status\":\"ok\",\"note\":\"log dice 'checkout ok' sin contexto util.\"}";
    }

    /** Observable: correlation ID propagado, log estructurado JSON, /logs endpoint. */
    private static String checkoutObservable(String totalRaw) {
        String corrId = UUID.randomUUID().toString();
        CTX.set(new RequestContext(corrId, "checkout-observable", Instant.now().toString()));
        try {
            double total = parseDoubleOr(totalRaw, 100.0);
            structuredLog("info", "checkout_start", Map.of("total", String.valueOf(total)));
            if (total > 500) {
                structuredLog("error", "checkout_failed", Map.of(
                        "total", String.valueOf(total),
                        "reason", "exceeds_limit",
                        "limit", "500"));
                observableErrors.increment();
                return "{\"variant\":\"observable\",\"status\":\"error\",\"correlation_id\":\"" + corrId +
                        "\",\"reason\":\"exceeds_limit\",\"limit\":500,\"total\":" + total + "}";
            }
            structuredLog("info", "checkout_ok", Map.of("total", String.valueOf(total)));
            return "{\"variant\":\"observable\",\"status\":\"ok\",\"correlation_id\":\"" + corrId +
                    "\",\"total\":" + total + ",\"note\":\"correlation_id propagado en logs estructurados.\"}";
        } finally {
            CTX.remove();
        }
    }

    private static void structuredLog(String level, String event, Map<String, String> fields) {
        RequestContext c = CTX.get();
        StringBuilder sb = new StringBuilder(256);
        sb.append("{\"ts\":\"").append(Instant.now()).append('"');
        sb.append(",\"level\":\"").append(level).append('"');
        sb.append(",\"event\":\"").append(event).append('"');
        if (c != null) {
            sb.append(",\"correlation_id\":\"").append(c.corrId).append('"');
            sb.append(",\"route\":\"").append(c.route).append('"');
        }
        for (Map.Entry<String, String> e : fields.entrySet()) {
            sb.append(",\"").append(e.getKey()).append("\":\"").append(escape(e.getValue())).append('"');
        }
        sb.append('}');
        String line = sb.toString();
        synchronized (recentLogs) {
            recentLogs.addFirst(line);
            while (recentLogs.size() > MAX_LOG_ENTRIES) recentLogs.removeLast();
        }
    }

    private static String logsJson() {
        StringBuilder sb = new StringBuilder(1024);
        sb.append("{\"entries\":[");
        synchronized (recentLogs) {
            boolean first = true;
            for (String line : recentLogs) {
                if (!first) sb.append(',');
                sb.append(line);
                first = false;
            }
        }
        sb.append("],\"max_kept\":").append(MAX_LOG_ENTRIES).append('}');
        return sb.toString();
    }

    private static String diagnosticsJson() {
        return "{\"stack\":\"" + STACK + "\",\"case\":\"" + CASE_NAME + "\"," +
                "\"legacy\":{\"requests\":" + legacyRequests.sum() + ",\"errors\":" + legacyErrors.sum() +
                ",\"observability\":\"println sin correlation, sin contexto\"}," +
                "\"observable\":{\"requests\":" + observableRequests.sum() + ",\"errors\":" + observableErrors.sum() +
                ",\"observability\":\"log estructurado JSON con correlation_id, /logs endpoint\"}}";
    }

    private record RequestContext(String corrId, String route, String startedAt) {}

    private static double parseDoubleOr(String raw, double dflt) {
        try { return Double.parseDouble(raw); } catch (Exception e) { return dflt; }
    }

    private static String escape(String v) {
        return v == null ? "" : v.replace("\\", "\\\\").replace("\"", "\\\"");
    }

    private static Map<String, String> queryParams(String rawQuery) {
        Map<String, String> params = new HashMap<>();
        if (rawQuery == null || rawQuery.isBlank()) return params;
        for (String pair : rawQuery.split("&")) {
            String[] parts = pair.split("=", 2);
            String k = URLDecoder.decode(parts[0], StandardCharsets.UTF_8);
            String v = parts.length > 1 ? URLDecoder.decode(parts[1], StandardCharsets.UTF_8) : "";
            params.put(k, v);
        }
        return params;
    }
}

File: cases/03-poor-observability-and-useless-logs/go/app/main.go
// Caso 03 — Observabilidad deficiente y logs inutiles (stack Go 1.23).
//
// Legacy: log.Printf sin correlation, sin contexto. Errores opacos.
// Observable: log estructurado JSON con correlation_id propagado, mas /logs
// que devuelve los ultimos N eventos.
//
// Primitivas Go distintivas, y el contraste que este caso hace visible:
//
//   - context.Context es LA via idiomatica de Go para propagar valores con
//     alcance de request. No es un ThreadLocal (Java) ni un AsyncLocal (.NET):
//     el contexto viaja como PARAMETRO EXPLICITO. Eso tiene una consecuencia
//     concreta — una funcion que no recibe ctx no puede leer el correlation_id
//     por accidente, y el compilador lo hace evidente. En Java/.NET el contexto
//     es ambiente: funciona hasta que alguien salta de thread y se pierde en
//     silencio. Aca perderlo es un error de compilacion, no un bug de runtime.
//
//   - log/slog (stdlib desde Go 1.21) emite JSON estructurado sin libreria
//     externa. Es el unico stack del lab donde el logger estructurado viene en
//     la biblioteca estandar.
package main

import (
	"context"
	"crypto/rand"
	"encoding/hex"
	"encoding/json"
	"log"
	"log/slog"
	"net/http"
	"os"
	"strconv"
	"sync"
	"sync/atomic"
	"time"
)

const (
	caseName      = "03 - Observabilidad deficiente y logs inutiles"
	maxLogEntries = 200
)

var stack = envOr("APP_STACK", "Go 1.23")

var (
	legacyRequests     int64
	legacyErrors       int64
	observableRequests int64
	observableErrors   int64
)

// ---------- contexto de request ----------

// ctxKey es un tipo privado: nadie fuera de este paquete puede colisionar con
// esta clave en el Context. Es la convencion Go para valores de contexto.
type ctxKey struct{}

type requestContext struct {
	CorrelationID string
	Route         string
	StartedAt     string
}

func withRequestContext(parent context.Context, rc requestContext) context.Context {
	return context.WithValue(parent, ctxKey{}, rc)
}

func fromContext(ctx context.Context) (requestContext, bool) {
	rc, ok := ctx.Value(ctxKey{}).(requestContext)
	return rc, ok
}

// ---------- buffer de logs estructurados ----------

type logEntry map[string]any

var (
	logsMu     sync.Mutex
	recentLogs []logEntry
)

func pushLog(e logEntry) {
	logsMu.Lock()
	defer logsMu.Unlock()
	recentLogs = append([]logEntry{e}, recentLogs...)
	if len(recentLogs) > maxLogEntries {
		recentLogs = recentLogs[:maxLogEntries]
	}
}

func snapshotLogs() []logEntry {
	logsMu.Lock()
	defer logsMu.Unlock()
	return append([]logEntry(nil), recentLogs...)
}

// structuredLog toma ctx como primer parametro — la firma obliga a pasarlo.
// Si el llamador no tiene contexto, no puede fingir que si.
func structuredLog(ctx context.Context, level, event string, fields map[string]any) {
	e := logEntry{
		"ts":    time.Now().UTC().Format(time.RFC3339Nano),
		"level": level,
		"event": event,
	}
	if rc, ok := fromContext(ctx); ok {
		e["correlation_id"] = rc.CorrelationID
		e["route"] = rc.Route
	}
	for k, v := range fields {
		e[k] = v
	}
	pushLog(e)

	// slog emite el mismo evento a stdout en JSON — el operador ve lo mismo
	// que /logs devuelve, sin tener que elegir entre uno u otro.
	attrs := []any{"event", event}
	if rc, ok := fromContext(ctx); ok {
		attrs = append(attrs, "correlation_id", rc.CorrelationID, "route", rc.Route)
	}
	for k, v := range fields {
		attrs = append(attrs, k, v)
	}
	slog.Info(event, attrs...)
}

// ---------- arranque ----------

func main() {
	slog.SetDefault(slog.New(slog.NewJSONHandler(os.Stdout, nil)))

	mux := http.NewServeMux()
	mux.HandleFunc("/", route)

	port := envOr("PORT", "8080")
	log.Printf("[case03-go] listening on %s", port)
	if err := http.ListenAndServe(":"+port, mux); err != nil {
		log.Fatalf("listen: %v", err)
	}
}

// ---------- routing ----------

func route(w http.ResponseWriter, r *http.Request) {
	path := r.URL.Path
	totalRaw := r.URL.Query().Get("total")

	status := http.StatusOK
	var payload any

	switch path {
	case "/", "/index":
		payload = map[string]any{
			"case":  caseName,
			"stack": stack,
			"routes": []string{
				"/health", "/checkout-legacy?total=100", "/checkout-observable?total=100",
				"/logs", "/metrics", "/diagnostics/summary", "/reset-lab",
			},
		}
	case "/health":
		payload = map[string]string{"status": "ok", "stack": stack, "case": caseName}
	case "/checkout-legacy":
		payload = checkoutLegacy(totalRaw)
		atomic.AddInt64(&legacyRequests, 1)
	case "/checkout-observable":
		payload = checkoutObservable(r.Context(), totalRaw)
		atomic.AddInt64(&observableRequests, 1)
	case "/logs":
		payload = map[string]any{"entries": snapshotLogs(), "max_kept": maxLogEntries}
	case "/metrics", "/diagnostics/summary":
		payload = diagnostics()
	case "/reset-lab":
		atomic.StoreInt64(&legacyRequests, 0)
		atomic.StoreInt64(&legacyErrors, 0)
		atomic.StoreInt64(&observableRequests, 0)
		atomic.StoreInt64(&observableErrors, 0)
		logsMu.Lock()
		recentLogs = nil
		logsMu.Unlock()
		payload = map[string]string{"status": "reset"}
	default:
		status = http.StatusNotFound
		payload = map[string]string{"error": "not_found", "path": path}
	}

	sendJSON(w, status, payload)
}

// ---------- endpoints ----------

// checkoutLegacy: log opaco. No recibe ctx — y esa es justamente la señal: no
// tiene forma de correlacionar nada aunque quisiera.
func checkoutLegacy(totalRaw string) map[string]any {
	total := parseFloatOr(totalRaw, 100.0)
	log.Printf("[INFO] processing checkout")
	if total > 500 {
		log.Printf("[ERROR] checkout failed")
		atomic.AddInt64(&legacyErrors, 1)
		return map[string]any{
			"variant": "legacy",
			"status":  "error",
			"note":    "log dice 'checkout failed' sin id, sin total, sin causa.",
		}
	}
	log.Printf("[INFO] checkout ok")
	return map[string]any{
		"variant": "legacy",
		"status":  "ok",
		"note":    "log dice 'checkout ok' sin contexto util.",
	}
}

// checkoutObservable: correlation ID en el Context, propagado explicitamente a
// cada llamada que loguea.
func checkoutObservable(parent context.Context, totalRaw string) map[string]any {
	corrID := newCorrelationID()
	ctx := withRequestContext(parent, requestContext{
		CorrelationID: corrID,
		Route:         "checkout-observable",
		StartedAt:     time.Now().UTC().Format(time.RFC3339Nano),
	})

	total := parseFloatOr(totalRaw, 100.0)
	structuredLog(ctx, "info", "checkout_start", map[string]any{"total": total})

	if total > 500 {
		structuredLog(ctx, "error", "checkout_failed", map[string]any{
			"total": total, "reason": "exceeds_limit", "limit": 500,
		})
		atomic.AddInt64(&observableErrors, 1)
		return map[string]any{
			"variant":        "observable",
			"status":         "error",
			"correlation_id": corrID,
			"reason":         "exceeds_limit",
			"limit":          500,
			"total":          total,
		}
	}
	structuredLog(ctx, "info", "checkout_ok", map[string]any{"total": total})
	return map[string]any{
		"variant":        "observable",
		"status":         "ok",
		"correlation_id": corrID,
		"total":          total,
		"note":           "correlation_id propagado via context.Context en logs estructurados.",
	}
}

func diagnostics() map[string]any {
	return map[string]any{
		"stack": stack,
		"case":  caseName,
		"legacy": map[string]any{
			"requests":      atomic.LoadInt64(&legacyRequests),
			"errors":        atomic.LoadInt64(&legacyErrors),
			"observability": "log.Printf sin correlation, sin contexto",
		},
		"observable": map[string]any{
			"requests":      atomic.LoadInt64(&observableRequests),
			"errors":        atomic.LoadInt64(&observableErrors),
			"observability": "log/slog estructurado con correlation_id via context.Context, /logs endpoint",
		},
	}
}

// ---------- helpers ----------

func newCorrelationID() string {
	b := make([]byte, 16)
	if _, err := rand.Read(b); err != nil {
		return strconv.FormatInt(time.Now().UnixNano(), 16)
	}
	return hex.EncodeToString(b)
}

func sendJSON(w http.ResponseWriter, status int, payload any) {
	body, err := json.Marshal(payload)
	if err != nil {
		http.Error(w, `{"error":"marshal"}`, http.StatusInternalServerError)
		return
	}
	w.Header().Set("Content-Type", "application/json; charset=utf-8")
	w.WriteHeader(status)
	_, _ = w.Write(body)
}

func envOr(key, fallback string) string {
	if v := os.Getenv(key); v != "" {
		return v
	}
	return fallback
}

func parseFloatOr(raw string, fallback float64) float64 {
	v, err := strconv.ParseFloat(raw, 64)
	if err != nil {
		return fallback
	}
	return v
}
